crypto/tls: Move draft-ietf-tls-esni-08 to -09

cjpatton opened the pull request:
Based on https://github.com/cloudflare/go/pull/46.

Spec changes
* (https://github.com/tlswg/draft-ietf-tls-esni/pull/355) Bump HPKE-05 to -07.
* (https://github.com/tlswg/draft-ietf-tls-esni/pull/353) Derive acceptance confirmation from handshake secret.
* (https://github.com/tlswg/draft-ietf-tls-esni/pull/352) Reuse HPKE context across HRR.
* (https://github.com/tlswg/draft-ietf-tls-esni/pull/360) Use a new code point to differentiate CHI from CHO.
* (https://github.com/tlswg/draft-ietf-tls-esni/pull/362) Bind context handle to encrypted payload.
* Bump extension code point.
* Revise the configuration identifier.

Other changes
* Remove `hrrPsk` from `ECHProvider.GetContext()` (API change).
* Prune retry configs of unknown version returned by the ECH provider.
* Add `EXP_ECHKeySet`, a default implementation of the ECH provider. (This will be useful for interop testing.)
* Require that the ECH extension not appear in OuterExtensions.
* Add event handler for outer SNI / public name mismatch.

wbl commented on src/crypto/tls/ech_provider.go:
```diff
@@ -57,22 +68,239 @@ type ECHProviderResult struct {
 }
+
+// EXP_ECHKeySet implements the ECHProvider interface for a sequence of ECH keys.
+type EXP_ECHKeySet struct {
```
The temptation is to go ahead and update while this is in use, with consequent terrible problems. Either this should be documented as something you shouldn't do, or there should be locks.

cjpatton replied:
> The temptation is to go ahead and update while this is in use, with consequent terrible problems.

I don't know what you mean here. Can you elaborate?

Added the following line to the comment: "NOTE: This API is EXPERIMENTAL and subject to change."

wbl replied:
so EXP_ECHKeySet is something a user would  pass into a tls configuration. But then the user wants to update it to implement key rotation, and it isn't threadsafe.

cjpatton replied:
I see. EXP_ECHKeySet is only meant to provide basic functionality. In particular, the set of supported ECH keys doesn't change over the lifetime of the TLS config. Should I document this?

wbl reviewed (commented):
This change is really messy because all the hpke stuff is intertwined with other bits. is there some way to rebase to make reviewing easier?

cjpatton commented on src/crypto/tls/handshake_messages.go:
```diff
@@ -123,15 +124,25 @@ func (m *clientHelloMsg) marshal() []byte {
-			if m.encryptedClientHelloOffered {
-				// draft-ietf-tls-esni-08, "encrypted_client_hello"
+			if len(m.esni) > 0 {
+				// draft-ietf-tls-esni-02, "encrypted_server_name"
```
Only as far as this: If the server detects this extension, then it aborts with "illegal_parameter". This is the server's behavior when the ESNI extension is malformed or uses a public key the server doesn't recognize.

cjpatton replied:
ESNI has been removed from this PR.

cjpatton commented on src/crypto/tls/ech_provider.go:
```diff
@@ -25,9 +34,11 @@ type ECHProvider interface {
+	// the handshake.  Note that, in some cases, it's appropriate to reject
+	// rather than abort. In particular, aborting with "illegal_parameter" might
+	// "stick out". res.Alert and res.Error are set.
+	GetContext(handle []byte, version uint16) (res ECHProviderResult)
```
Do yo have a suggestion for an alternative? Perhaps `GetHPKEOpener`?

wbl replied:
I think that name is better.

cjpatton replied:
@chris-wood and @Lekensteyn, let me know if you have a strong preference about the name. I'm fine changing it.

Lekensteyn replied:
Maybe `GetDecryptionContext`? I think that Watson was concerned with `GetContext` appearing related to [context](https://golang.org/pkg/context/) which is not the case.

cjpatton replied:
I'd be fine with that I suppose.

cjpatton replied:
Done.

cjpatton commented:
@wbl, focus your review on the last commit, which now deals with the TLS changes only.

wbl reviewed (approved):
One comment about names. I'm leaving the HPKE stuff for another review.

Lekensteyn reviewed (changes requested):
Unless there is a reason to keep ESNI, please drop it for simplicity.
I think there is also a bug in the calculation of a config ID.

(Observation: `echKey` was moved from tests to the public `EXP_ECHKey` type. Presumably to reduce code duplication in its users.)

I skimmed through the rest and did not notice other issues.

cjpatton commented on src/crypto/tls/ech_config.go:
```diff
@@ -174,10 +157,35 @@ func (config *ECHConfig) isPeerCipherSuiteSupported(suite echCipherSuite) bool {
+
+	id := make([]byte, 8)
+	n, err := hkdf.Expand(h.New,
+		hkdf.Extract(h.New, nil, config.raw),
```
Yup, nice catch!

cjpatton commented on src/crypto/tls/ech_provider.go:
```diff
@@ -57,22 +68,241 @@ type ECHProviderResult struct {
+		for _, suite := range key.config.suites {
+			if !hpke.KdfID(suite.kdfId).IsValid() ||
+				!hpke.AeadID(suite.aeadId).IsValid() {
+				skip = true
```
THANK YOU. I was wondering about the correct goto syntax.

Done, here and in ech_config.go.

cjpatton commented on src/crypto/tls/esni.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// Generates a well-formed ESNI extension with random values. (See
+// draft-ietf-tls-esni-02.)
```
Ack, will drop from this PR and bring this up again in a follow-up PR.

chris-wood commented on src/crypto/tls/ech_config.go:
```diff
@@ -174,10 +158,35 @@ func (config *ECHConfig) isPeerCipherSuiteSupported(suite echCipherSuite) bool {
 }
+
+// id returns the configuration identifier for the given KDF.
+func (config *ECHConfig) id(kdfId hpke.KdfID) ([]byte, error) {
```
This should use the HPKE KDF API directly, after [@armfazh's PR](https://github.com/armfazh/circl/pull/8) lands.

cjpatton replied:
Done.

cjpatton commented:
> Unless there is a reason to keep ESNI, please drop it for simplicity.
> I think there is also a bug in the calculation of a config ID.

Dropped, by the way :+1:

Lekensteyn commented on src/crypto/tls/handshake_client_tls13.go:
```diff
@@ -815,7 +826,7 @@ func (hs *clientHandshakeStateTLS13) abortIfRequired() error {
 		// If ECH was rejected, then abort the handshake.
 		c.sendAlert(alertECHRequired)
-		return errors.New("tls: ech: rejected")
+		return errors.New("tls: ech: rejected" + string(base64.StdEncoding.EncodeToString(c.ech.retryConfigs)))
```
`string(...)` is redundant as `EncodeToString` already returns one. Alternatively:
```suggestion
		return fmt.Errorf("tls: ech: rejected %s", base64.StdEncoding.EncodeToString(c.ech.retryConfigs))`
```
Is it really important to log this or can this addition be dropped?

cjpatton replied:
Deja Vu :)

This was done for testing and wasn't meant to be merged. Reverted.

Lekensteyn reviewed (commented):
Looks mostly fine, only minor typos and two parsing issues.

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
 
-	// Produce covertext for the ECH extension.
+	// Ensure that only TLS 1.3 and above are offered.
+	if v := helloInner.supportedVersions; v[len(v)-1] < VersionTLS13 {
```
Yup, this always ought to be the case. But to play it safe, I changed the condition to
```
len(v) == 0 || v[len(v)-1] < VersionTLS13
```

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -643,15 +479,60 @@ func echReadContextHandle(s *cryptobyte.String, handle *echContextHandle) bool {
+}
+
+// echEncodeClientHelloInner interprets ineerData as a ClientHelloInner message
+// and transforms it into an EncodedClinetHelloInner. It also returns a bool
```
Good idea. Fixed here and below.

cjpatton commented on src/crypto/tls/ech_config.go:
```diff
@@ -33,60 +30,59 @@ type ECHConfig struct {
 		}
+		config.raw = raw[:n]
+		configs = append(configs, config)
 		raw = raw[n:]
```
Done. Great catch!

cjpatton commented on src/crypto/tls/ech_config.go:
```diff
@@ -33,60 +30,59 @@ type ECHConfig struct {
+	return b.Bytes()
 }
 
 func readConfigContents(contents *cryptobyte.String, config *ECHConfig) bool {
```
Done.

cjpatton commented:
Great catches, thanks!

Lekensteyn reviewed (approved):
LGTM

cjpatton commented:
Rebased (now using HPKE-07!).

cjpatton commented:
Rebased.

chris-wood commented on src/crypto/tls/ech_test.go:
```diff
@@ -1016,3 +813,39 @@ func TestECHHandshake(t *testing.T) {
 	}
 }
+
+func TestUnmarshalConfigs(t *testing.T) {
```
Do we have tests for the handle marshal/unmarshal logic? (It's trivial, but, you know.)

cjpatton replied:
Yes, see the end of ech_test.go. They could be more coimplete, I think ... @Lekensteyn recently found a bug here, so it's a good question.

chris-wood replied:
👍 Cool beans

chris-wood commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
+	}
+
+	if hello.echIsInner {
+		if hello.echIsOuter {
```
Do we have a check somewhere for the opposite case? Namely, if this is CHOuter and "ech_is_inner" is present?

cjpatton replied:
Yes, see the new test case in TestECHHandshake.

chris-wood replied:
👍

chris-wood reviewed (commented):
This looks great! As always, it's a pleasure reading your PRs. I went through each of the issues and they appear to be correctly implemented. My only substantial comment (and it's really kind of a nit) is around how and when we set the "greased" bit. It seems that it's always (?) set upon rejection, but it's not clear to me that this is correct. (It may be the case that we really don't need to distinguish rejection due to GREASE and rejection due to decryption failure, but it's worth asking the question, I think.)

It would be nice if we had tests for the following cases (if they aren't already there):

- CHInner contains "encrypted_client_hello" and "ech_is_inner" extensions.
- CHOuter contains a `config_id` corresponding to a legitimate ECHConfig but with an incorrect length, e.g., 4 bytes (too short) or 12 bytes (too long). 
- HRR triggering with non-zero `config_id` and `enc` values in the CHOuter2.

It would also be good if we had explicit tests for `EXP_ECHKeySet`.

Maybe these tests are already here and I missed them, or they're already covered by the existing default tests, in which case feel free to ignore this. :-)

cjpatton commented on src/crypto/tls/ech_provider.go:
```diff
@@ -1,22 +1,34 @@
+)
+
+const (
+	maxConfigIdLen = 255
```
That's part of the idea, yeah.

chris-wood replied:
👍

cjpatton commented on src/crypto/tls/ech_provider.go:
```diff
@@ -57,22 +71,242 @@ type ECHProviderResult struct {
+//     ECHConfig config<0..2^16>; // draft-ietf-tls-esni-09
+// } ECHKey;
+type EXP_ECHKey struct {
+	config ECHConfig
+	sk     kem.PrivateKey
```
Done.

cjpatton commented on src/crypto/tls/ech_provider.go:
```diff
@@ -57,22 +71,242 @@ type ECHProviderResult struct {
+// that the ECH standard does not specify this format.
+//
+// struct {
+//     opaque private_key<0..2^16-1>;
```
Done.

cjpatton commented on src/crypto/tls/handshake_client_tls13.go:
```diff
@@ -462,6 +441,37 @@ func (hs *clientHandshakeStateTLS13) establishHandshakeKeys() error {
+			c.sendAlert(alertInternalError)
+			return errors.New("tls: internal error: failed to clone hash")
+		}
+		var random [32]byte
```
Done.

cjpatton commented on src/crypto/tls/handshake_server_tls13.go:
```diff
@@ -597,13 +607,26 @@ func illegalClientHelloChange(ch, ch1 *clientHelloMsg) bool {
+			c.sendAlert(alertInternalError)
+			return errors.New("tls: internal error: failed to clone hash")
+		}
+		confTranscript.Write(hs.clientHello.marshal())
```
What do you mean by "raw representation"? If you mean: should the input to he hash begin with the ClientHello (i.e., `hs.clientHello.marshal()`), it depends on whether the HRR path has been taken. If it has, then starting with a fresh transcript would be incorrect. To prove this to yourself, replace:
```
confTranscript := cloneHash(hs.transcript, hs.suite.hash)
```
with
```
confTranscript := hs.suite.hash.New() 
```
This will cause the client to issue a "bad record MAC" alert in one of the ECH unit tests.

chris-wood replied:
Sorry, I meant should `hs.clientHello.raw` be `nil` before calling `marshal` on it, not whether or not the transcript is reset. (I agree that it shouldn't be for the reason you state.)

chris-wood replied:
(Chatted offline, since there's no CH rewriting)

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -83,20 +78,24 @@ func (e EXP_EventECHServerStatus) Name() string {
-	81, 163, 222, 52, 211, 54, 210, 46, 37, 78, 216, 157, 97, 241, 244,
+// EXP_EventECHPublicNameMismatch is emitted if the outer SNI does not match
+// match the public name of the ECH configuration. Note that we do not record
+// the outer SNI in order to avoid collecting this potentially sensitive data.
```
Ack, leaving as is for now. It would be a relatively simple change to make in the future, should we need it.

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
-		}
+	// Ask the ECH provider for the HPKE context.
+	if c.ech.opener == nil {
+		res := p.GetDecryptionContext(ech.handle.marshal(), extensionECH)
```
Yup! It simply interprets the config_id as a uint8-prefixed string and checks if it matches the id of an existing config. Note that there's a test for this in TestECHProvider.

chris-wood replied:
👍 Should we note that in the API?

cjpatton replied:
What would suggest? I don't think it's necessary, since this behavior is spec compliant.

chris-wood replied:
Oh, I was just thinking adding a comment. But we don't need to. Food for thought!

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
+			}
+		case ECHProviderReject:
+			// Reject ECH.
+			c.ech.greased = true
```
We do not know at this point whether the client intended to offer or grease ECH, so we presume grease until it sends an "ech_required" alert. I added a comment explaining this, here and below.

cjpatton commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,287 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
-
 		// Reject ECH.
-		return reject()
+		c.ech.greased = true
```
Resolved above.

cjpatton commented:
> My only substantial comment (and it's really kind of a nit) is around how and when we set the "greased" bit. It seems that it's always (?) set upon rejection, but it's not clear to me that this is correct. (It may be the case that we really don't need to distinguish rejection due to GREASE and rejection due to decryption failure, but it's worth asking the question, I think.)

Recall that, if the server doesn't recognize the config_id, then it has know way of knowing whether the client intended to offer or grease ECH. To resolve this, we _presume_ grease until the client sends an "ech_required", which indicates that it intended to offer ECH. (I had a comment in `echAccepOrReject` explaining this before, but I deleted it. I added the comment back in.)

> It would be nice if we had tests for the following cases (if they aren't already there):
> 
>     * CHInner contains "encrypted_client_hello" and "ech_is_inner" extensions.
I added this case to TestECHHandshake.

>     * CHOuter contains a `config_id` corresponding to a legitimate ECHConfig but with an incorrect length, e.g., 4 bytes (too short) or 12 bytes (too long).

I would argue that a `config_id` that is truncated too short or too long isn't a legitimate ID by definition. In any case, I added a test case for a short `config_id` to TestECHProvider (new in ech_test.go).

>     * HRR triggering with non-zero `config_id` and `enc` values in the CHOuter2.

I modified the `testingECHIllegalHandleAfterHRR` to send a non-empty `config_id` and `enc`. (See handshake_client_tls13.go.)

> 
> It would also be good if we had explicit tests for `EXP_ECHKeySet`.

I added a few test cases to TestECHProvider.

chris-wood commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,294 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
+			}
+		case ECHProviderReject:
+			// Reject ECH. We do not know at this point whether the client
+			// intended to offer or grease ECH, so we presume grease until the c
```
```suggestion
			// intended to offer or grease ECH, so we presume grease until the
```

chris-wood commented on src/crypto/tls/ech.go:
```diff
@@ -105,327 +104,294 @@ func (c *Conn) echOfferOrGrease(helloBase *clientHelloMsg) (hello, helloInner *c
-		// Reject ECH.
-		return reject()
+		// Reject ECH. We do not know at this point whether the client
+		// intended to offer or grease ECH, so we presume grease until the c
```
```suggestion
		// intended to offer or grease ECH, so we presume grease until the
```

chris-wood reviewed (commented):
Couple of nits!

cjpatton commented:
Squashed.
